test/e2e_node/container_lifecycle_test.go

```diff
@@ -3568,7 +4289,415 @@ var _ = SIGDescribe(nodefeature.SidecarContainers, "Containers Lifecycle", func(
+		init1 := "init-1"
+		restartableInit1 := "restartable-init-1"
+		init2 := "init-2"
+		restartableInit2 := "restartable-init-2"
+		regular1 := "regular-1"
+
+		containerTerminationSeconds := 5
+		updatedImage := busyboxImage
+
+		var originalPodSpec *v1.Pod
+
+		ginkgo.BeforeEach(func(ctx context.Context) {
+			originalPodSpec = &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "update-restartable-init-img-during-initialize",
+				},
+				Spec: v1.PodSpec{
+					RestartPolicy: v1.RestartPolicyAlways,
+					InitContainers: []v1.Container{
+						{
+							Name:  init1,
+							Image: agnhostImage,
+							Command: ExecCommand(init1, execCommand{
+								Delay:    1,
+								ExitCode: 0,
+							}),
+						},
+						{
+							Name:  restartableInit1,
+							Image: agnhostImage,
+							Command: ExecCommand(restartableInit1, execCommand{
+								Delay:              600,
+								TerminationSeconds: containerTerminationSeconds,
+								ExitCode:           0,
+							}),
+							RestartPolicy: &containerRestartPolicyAlways,
+						},
+						{
+							Name:  init2,
+							Image: agnhostImage,
+							Command: ExecCommand(init2, execCommand{
+								Delay:    600,
+								ExitCode: 0,
+							}),
+						},
+						{
+							Name:  restartableInit2,
+							Image: agnhostImage,
+							Command: ExecCommand(restartableInit2, execCommand{
+								Delay:              1,
+								TerminationSeconds: 1,
+								ExitCode:           0,
+							}),
+							RestartPolicy: &containerRestartPolicyAlways,
+						},
+					},
+					Containers: []v1.Container{
+						{
+							Name:  regular1,
+							Image: agnhostImage,
+							Command: ExecCommand(regular1, execCommand{
+								Delay:              1,
+								TerminationSeconds: 1,
+								ExitCode:           0,
+							}),
+						},
+					},
+				},
+			}
+		})
+
+		ginkgo.It("Should begin initializing the pod, restartPolicy=Always", func(ctx context.Context) {
+			preparePod(originalPodSpec)
+
+			client := e2epod.NewPodClient(f)
+			pod := client.Create(ctx, originalPodSpec)
+
+			ginkgo.By("Running the pod", func() {
+				err := e2epod.WaitForPodCondition(ctx, f.ClientSet, pod.Namespace, pod.Name, "the second init container is running but not started", 2*time.Minute, func(pod *v1.Pod) (bool, error) {
+					if pod.Status.Phase != v1.PodPending {
+						return false, fmt.Errorf("pod should be in pending phase")
+					}
+					if len(pod.Status.InitContainerStatuses) < 3 {
+						return false, nil
+					}
+					containerStatus := pod.Status.InitContainerStatuses[2]
+					return containerStatus.State.Running != nil && *containerStatus.Started, nil
+				})
+				framework.ExpectNoError(err)
+			})
+
+			ginkgo.By("Changing the image of the restartable init container", func() {
+				client.Update(ctx, pod.Name, func(pod *v1.Pod) {
+					pod.Spec.InitContainers[1].Image = updatedImage
+				})
+			})
+
+			//TODO verify this is the intended behavior
```

Comment:
This is an interesting case. When the initialization looks like init -> sidecar -> long-running init, when the sidecar image is updated, should the sidecar restart? From a design perspective, it's simpler to just define that whenever the image changes, a running sidecar always restarts. But from the perspective of a user running a complex initialization, having the sidecar restart mid-initialization may not be desired.

Reply:
I believe we have the same behavior if sidecar will crash. So the Pod should expect any of sidecars to restart while the initialization is running.